### ISSUE TITLE
Can't run on GPU with python 3.8

### ISSUE THREAD

**Issue by mafortin** [BBillot/SynthSeg#49]
Hi! 

I've been using SynthSeg for a few months now on python 3.6 and wanted to test on python 3.8. No problem at all with the python 3.6 configuration. However, after creating a new virtual python environment with the 'requirements_python38.txt' file and manually installing cudatoolkit=10.0 and cudnn=7.6, SynthSeg never detects my GPU resources, so it runs on CPU only. 

I've been trying several different solutions but nothing ended up working.

The explicit steps I'm doing (in my terminal [I'm using Ubuntu 18.04) are the ones below:

1) 'conda create -n synthseg38 python=3.8'
2) 'conda activate synthseg38'
3) 'pip install -r /path/to/SynthSeg/requirements_python3.8.txt'
4) 'conda install -c conda-forge cudatoolkit=10.0 cudnn=7.6'

I get the following 'error' messages (still runs but on CPU) when I try to run SynthSeg with GPU:

""" 
Using TensorFlow backend.

SynthSeg-robust 2.0

using 1 thread
2023-02-28 10:48:10.895447: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2023-02-28 10:48:11.182097: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:981] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2023-02-28 10:48:11.182307: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties: 
pciBusID: 0000:00:06.0 name: Tesla P100-PCIE-16GB computeCapability: 6.0
coreClock: 1.3285GHz coreCount: 56 deviceMemorySize: 15.90GiB deviceMemoryBandwidth: 681.88GiB/s
2023-02-28 10:48:11.182553: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory
2023-02-28 10:48:11.182648: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcublas.so.10'; dlerror: libcublas.so.10: cannot open shared object file: No such file or directory
2023-02-28 10:48:11.217937: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2023-02-28 10:48:11.218516: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2023-02-28 10:48:11.218643: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusolver.so.10'; dlerror: libcusolver.so.10: cannot open shared object file: No such file or directory
2023-02-28 10:48:11.218721: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcusparse.so.10'; dlerror: libcusparse.so.10: cannot open shared object file: No such file or directory
2023-02-28 10:48:11.223702: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2023-02-28 10:48:11.223741: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1598] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.
Skipping registering GPU devices...
2023-02-28 10:48:11.224138: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2023-02-28 10:48:11.232404: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 2095074999 Hz
2023-02-28 10:48:11.232595: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7fe2f8000b20 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2023-02-28 10:48:11.232625: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2023-02-28 10:48:11.234544: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2023-02-28 10:48:11.234573: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      
predicting 1/1
"""

I am expecting this to be either an error in librairies compatibility or something missing from the requirements.txt file for python 3.8. I am not an expert at all in tensorflow/cuda packages, so sorry if this is irrelevant, but I would appreciate some guidances! :)

**Comment by Pavelevich** [comment]
Maybe the version of the CUDATOOLKIT package that you are installing is not compatible with your version of Python. The CUDATOOLKIT=10.0 package that you are installing is designed to work with Python 3.7, and may not be compatible with Python 3.8.

To fix this, you can try installing a different version of cudatoolkit that is compatible with Python 3.8. 
For example, you can try installing CUDATOOLKIT=11.0, which is designed to work with Python 3.8:
conda install -c conda-forge cudatoolkit=11.0 cudnn=8.0

**Comment by mafortin** [comment]
Hi @Pavelevich ,

thanks for the clarification and suggestion! :) However, it didn't solve the problem...

@BBillot: Even if this is not the solution, it's still relevant to mention the difference in cudatoolkit version required to work with python 3.8 in the documentation. :)

**Comment by Pavelevich** [comment]
> Hi @Pavelevich ,
> 
> thanks for the clarification and suggestion! :) However, it didn't solve the problem...
> 
> @BBillot: Even if this is not the solution, it's still relevant to mention the difference in cudatoolkit version required to work with python 3.8 in the documentation. :)

hmmm ok I think u can try this too:

1. _Check that the NVIDIA drivers are correctly installed and up-to-date. You can do this by running the nvidia-smi command in the terminal. If the command does not run or does not show any information, the drivers may not be installed correctly._

2. _Verify that the virtual environment is activated when running SynthSeg. You can do this by checking the environment variable $CONDA_PREFIX. If the variable is not set or does not point to your virtual environment, the environment may not be activated correctly._

3. _Try running other GPU-accelerated applications to confirm that your GPU is working properly. If other applications are also not detecting the GPU, the issue may be with the GPU itself or the system configuration._

If none of the above steps work, you can try running SynthSeg with the CUDA_VISIBLE_DEVICES environment variable set to the index of your GPU. 
**if your GPU is index 0** then
 you can run CUDA_VISIBLE_DEVICES=0 python script.py to force the script to use the GPU.

**Comment by mafortin** [comment]
1) This prints information about my GPU, which I'd expect to be the correct output. 
    - However, it says CUDA Version: 11.6 in the top right corner. Considering I have installed 11.0 with your previous proposition, could it be this creating the issue?
2) It gives the correct directory to my conda virtual environment.
3) SynthSeg runs perfectly on the same GPU I'm trying to access with python 3.6 on a different terminal window.

**Comment by Pavelevich** [comment]
> * However, it says CUDA Version: 11.6 in the top right corner. Considering I have installed 11.0 with your previous proposition, could it be this creating the issue

yes, maybe.

**Comment by mafortin** [comment]
Even with cudatoolkit=11.6 and cudnn=8.4 it didn't work (same error message)

**Comment by BBillot** [comment]
Hello both,

thanks for the suggestions :)

I actually never tried to run SynthSeg on the GPU with python3.8, because my good old python3.6 was working very well. 

I added the python 3.8 libraries after several people requested this, because 3.6 is no longer maintained. To be honest, it was a bit of a headache to find compatible versions of all the libraries (numpy, tensorflow, keras, etc.) for python 3.8. Eventually I found versions that run smoothly, but incompatibilities are still detected when installing the packages. So this means that you might have to re-install some packages several times, until they exactly match the list in requirements_python38.txt. I know this is silly, but that's a short-term solution...

So basically, make sure that the list of packages you have installed exactly match the requirements. Let me know if that fixes the issue. I'm guessing it won't, but at least we have comparable set-up.

**Comment by mafortin** [comment]
Hi @BBillot ,

ok, thanks for letting us know! :) 

However, could you at least mention in the documentation (in the installation section) that it requires a different version of cudatoolkit/cudnn for the python 3.8 version?

**Comment by BBillot** [comment]
Okay, I will add this to the README :slightly_smiling_face:
let me know if the above fixes your problem

**Comment by BBillot** [comment]
Hello both. I investigated to find a stable version for python 3.8 and the new requirements are listed in requirements_python3.8.txt. The new version has way fewer libraries because I've been doing some cleaning in the ext/ folder. 

The new version uses tensorflow 2.2.0 and keras 2.3.1. Importantly, to make this run on the GPU it needs cuda 10.1 and cudnn 7.6.5.

I added minimal pip/conda commands in the README to install the required libraries.

Finally, I tested these new environments and I was able to make then run on the GPU, even for Python 3.8.

PS: I also did all the above for python 3.6, just to stay consistent.
